### DISCUSSION TITLE
How to build multiple binaries, with multiple and differents go files

### DISCUSSION THREAD

**Discussion by sebbbastien** [goreleaser/goreleaser#1531] [Q&A] [answered]
Hi,

I have to build two binaries, one produce data and store them in a database, the other consume data from the database. These two binaries share the connection information to the database, and they must be standalone files, I can't use a side config file (json/ini, etc.) to store database credentials.

I need to provide this pair of binaries for multiple clients, so they must not use the same database IP/login/password information.

I have two main files, mainProducer.go and mainConsumer.go. They declare main package and main().
Alongside I have one Go file for each customer, envA.go, envB.go, etc, each file set a database connection struct, with the same fields, but differents values.

Without goreleaser I can run `go build mainProducer.go envA.go` and `go build mainConsumer.go envA.go` to build binaries for customer A.

I haven't figured out how to build two binaries with Goreleaser.

Do you see a solution using Goreleaser?

Best regards,

**Comment by caarlos0** [comment, marked as answer]
I think you can just put both files on the `main` property of builds: https://goreleaser.com/customization/build/

The environment setup you should probably make configurable, as you're probably embedding database secrets in your binaries.

https://github.com/caarlos0/env might help you make that easy :)

**Reply by sebbbastien** [reply]
Hi @caarlos0 

I'll try again with multiple files in `build/main`, but I think I've already try this.

> https://github.com/caarlos0/env might help you make that easy :)

Yes, your lib seams great, I think it will help me I think.

Thank you for your help

**Comment by sebbbastien** [comment]
Hi @caarlos0,

I've write some lines to test: https://github.com/sebbbastien/goProgram

In `subfolder/` I have `mainA.go` and `mainB.go`, both of them use `v` who is defined in `shared.go`.

From the command line I can have:
```
» cd subfolder/
» go run mainA.go shared.go
From mainA(), V = 'shared value with A and B'
» go run mainB.go shared.go
From mainB(), V = 'shared value with A and B'
```
```
» cd ../
» go run subfolder/mainA.go subfolder/shared.go
From mainA(), V = 'shared value with A and B'
» go run subfolder/mainB.go subfolder/shared.go
From mainB(), V = 'shared value with A and B'
```

I've try multiple way to compile this with goreleaser. From the "root" folder:
```
» cat .goreleaser_subfolder.yml
[...]
builds:
  - id: mainA
    env:
      - CGO_ENABLED=0
    dir: subfolder/
    main: mainA.go shared.go
[...]
```
The output is:
```
» goreleaser release --snapshot --rm-dist --config=.goreleaser_subfolder.yml
[...]
   ⨯ release failed after 0.02s error=stat subfolder/mainA.go shared.go: no such file or directory
```

I tried two other solutions, another from "root" directory:
```
» cat .goreleaser_relativepath.yml
[...]
builds:
  - id: mainA
    env:
      - CGO_ENABLED=0
    main: subfolder/mainA.go subfolder/shared.go
[...]
```
The output is:
```
» goreleaser release --snapshot --rm-dist --config=.goreleaser_relativepath.yml
[...]
   ⨯ release failed after 0.02s error=stat subfolder/mainA.go subfolder/shared.go: no such file or directory
```

And a last from the subfolder/
```
» cd subfolder/
cat .goreleaser.yml
[...]
builds:
  - id: mainA
    env:
      - CGO_ENABLED=0
    main: mainA.go shared.go
[...]
```
The output is:
```
» goreleaser release --snapshot --rm-dist --config=.goreleaser.yml
[...]
   ⨯ release failed after 0.03s error=stat mainA.go shared.go: no such file or directory
```

Do you see another way to do that with goreleaser?

**Reply by caarlos0** [reply]
you can use `shared` as a lib:

```
- shared.go
- cmd/a/main.go
- cmd/b/main.go
```

both `main.go` files then `import` the root pkg:

```go
package main

import "github.com/sebbbastien/goProgram"

// etc..
```

goreleaser then will have 2 build configs:

```yaml
builds:
- id: a
  main: ./cmd/a/main.go
- id: b
  main: ./cmd/b/main.go
```

and that should work.

the way you have your project structured its not possible to build in goreleaser right now.

**Reply by aaomidi** [reply]
This brings up an interesting point about potentially being able to use the `id` value of a build (or any other step) as a variable for templating. For example, the `binary` field in `builds[0]` could be `{{ .Id }}`.

**Reply by suntong** [reply]
> you can use `shared` as a lib:
> both `main.go` files then `import` the root pkg...

So I gave it a try, but get

```
error=failed to add: 'dist/mainB_windows_arm64/gorelease2.exe' -> 'gorelease2.exe': add gorelease2.exe: file already exists
58
Error: The process '/opt/hostedtoolcache/goreleaser-action/1.19.2/x64/goreleaser' failed with exit code 1
```

How to fix it please? The whole project is at https://github.com/suntong/gorelease2.

Thanks

**Reply by suntong** [reply]
How about this @caarlos0 ?

**Reply by caarlos0** [reply]
you need to set different binary names for each, otherwise they'll get the same name, and you'll get the error you're seeing

**Comment by lonnywong** [comment]
How to build multiple binaries with the same `id`?

**Reply by caarlos0** [reply]
If its the same build parameters, just set `goos` and `goarch` according to the documentation: https://goreleaser.com/customization/builds/

If you need different build flags for each platform or something like that, you'll have to set different IDs for each build.
